hudi-common/src/main/java/org/apache/hudi/common/util/queue/DisruptorExecutor.java

```diff
@@ -0,0 +1,135 @@
+ */
+public class DisruptorExecutor<I, O, E> extends HoodieExecutorBase<I, O, E> {
+
+  private static final Logger LOG = LogManager.getLogger(DisruptorExecutor.class);
+  private final HoodieMessageQueue<I, O> queue;
+
+  public DisruptorExecutor(final Option<Integer> bufferSize, final Iterator<I> inputItr,
+                           IteratorBasedQueueConsumer<O, E> consumer, Function<I, O> transformFunction, Option<String> waitStrategy, Runnable preExecuteRunnable) {
+    this(bufferSize, new IteratorBasedQueueProducer<>(inputItr), Option.of(consumer), transformFunction, waitStrategy, preExecuteRunnable);
+  }
+
+  public DisruptorExecutor(final Option<Integer> bufferSize, HoodieProducer<I> producer,
+                           Option<IteratorBasedQueueConsumer<O, E>> consumer, final Function<I, O> transformFunction) {
+    this(bufferSize, producer, consumer, transformFunction, Option.of(WaitStrategyFactory.DEFAULT_STRATEGY), Functions.noop());
+  }
+
+  public DisruptorExecutor(final Option<Integer> bufferSize, HoodieProducer<I> producer,
+                           Option<IteratorBasedQueueConsumer<O, E>> consumer, final Function<I, O> transformFunction, Option<String> waitStrategy, Runnable preExecuteRunnable) {
+    this(bufferSize, Collections.singletonList(producer), consumer, transformFunction, waitStrategy, preExecuteRunnable);
+  }
+
+  public DisruptorExecutor(final Option<Integer> bufferSize, List<HoodieProducer<I>> producers,
+                           Option<IteratorBasedQueueConsumer<O, E>> consumer, final Function<I, O> transformFunction,
+                           final Option<String> waitStrategy, Runnable preExecuteRunnable) {
+    super(producers, consumer, preExecuteRunnable);
+    this.queue = new DisruptorMessageQueue<>(bufferSize, transformFunction, waitStrategy, producers.size(), preExecuteRunnable);
+  }
+
+  /**
+   * Start all Producers.
+   */
+  @Override
+  public CompletableFuture<Void> startProducers() {
+    return CompletableFuture.allOf(producers.stream().map(producer -> {
+      return CompletableFuture.supplyAsync(() -> {
+        try {
+          producer.produce(queue);
+        } catch (Throwable e) {
+          LOG.error("error producing records", e);
+          throw new HoodieException("Error producing records in disruptor executor", e);
+        }
+        return true;
+      }, producerExecutorService);
+    }).toArray(CompletableFuture[]::new));
+  }
+
+  @Override
+  protected void setup() {
+    ((DisruptorMessageQueue)queue).setHandlers(consumer.get());
+    ((DisruptorMessageQueue)queue).start();
+  }
+
+  @Override
+  protected void postAction() {
+    try {
+      queue.close();
+      super.close();
+    } catch (IOException e) {
+      throw new HoodieIOException("Catch IOException while closing DisruptorMessageQueue", e);
+    }
+  }
+
+  @Override
+  protected CompletableFuture<E> startConsumer() {
+    return producerFuture.thenApplyAsync(res -> {
+      try {
+        queue.close();
```

Comment:
I see we are calling queue.close() from many places. can you confirm that its idempotent and will not cause any exceptions.

Reply:
Changed! Only close the queue if this queue is not closed before.